lib/auth/machineid/machineidv1/spiffe_federation_syncer.go

```diff
@@ -0,0 +1,629 @@
+		return "no_current_bundle"
+	}
+	if in.Status.CurrentBundleSyncedAt.AsTime().IsZero() {
+		log.DebugContext(ctx, "No status.current_bundle_synced_at, will sync")
+		return "no_current_bundle_synced_at"
+	}
+	// Check if we've passed the next sync time.
+	nextSyncAt := in.Status.NextSyncAt.AsTime()
+	now := clock.Now()
+	if !nextSyncAt.IsZero() && now.After(nextSyncAt) {
+		log.DebugContext(
+			ctx,
+			"status.next_sync_at has passed, will sync",
+			"next_sync_at", nextSyncAt,
+			"now", now,
+		)
+		return "next_sync_at_passed"
+	}
+	// Check to see if the configured bundle source has changed
+	if in.Status.CurrentBundleSyncedFrom != nil {
+		if !proto.Equal(in.Spec.BundleSource, in.Status.CurrentBundleSyncedFrom) {
+			log.DebugContext(ctx, "status.current_bundle_synced_from has changed, will sync")
+			return "bundle_source_changed"
+		}
+	}
+
+	return ""
+}
+
+func (s *SPIFFEFederationSyncer) syncTrustDomain(
+	ctx context.Context, name string,
+) (out *machineidv1.SPIFFEFederation, err error) {
+	ctx, span := tracer.Start(ctx, "SPIFFEFederationSyncer/syncTrustDomain")
+	defer func() {
+		tracing.EndSpan(span, err)
+	}()
+
+	current, err := s.cfg.Store.GetSPIFFEFederation(ctx, name)
+	if err != nil {
+		return nil, trace.Wrap(err, "failed to get SPIFFEFederation")
+	}
+
+	log := s.cfg.Logger.With(
+		"current_revision", current.GetMetadata().GetRevision(),
+		"trust_domain", current.GetMetadata().GetName(),
+	)
+
+	td, err := spiffeid.TrustDomainFromString(current.GetMetadata().GetName())
+	if err != nil {
+		return nil, trace.Wrap(err, "parsing metadata.name as trust domain name")
+	}
+
+	// Determine - should we sync...
+	syncReason := shouldSyncTrustDomain(ctx, log, s.cfg.Clock, current)
+	if syncReason == "" {
+		log.DebugContext(ctx, "Skipping sync as is not required")
+		return current, nil
+	}
+	log.InfoContext(ctx, "Sync starting", "reason", syncReason)
+
+	// Clone the persisted resource so we can compare to it.
+	out = proto.Clone(current).(*machineidv1.SPIFFEFederation)
+
+	// Refresh...
+	if out.Status == nil {
+		out.Status = &machineidv1.SPIFFEFederationStatus{}
+	}
+
+	var bundle *spiffebundle.Bundle
+	var nextSyncIn time.Duration
+	switch {
+	case current.Spec.BundleSource.HttpsWeb != nil:
+		url := current.Spec.BundleSource.HttpsWeb.BundleEndpointUrl
+		log.DebugContext(
+			ctx,
+			"Fetching bundle using https_web profile",
+			"url", url,
+		)
+		bundle, err = federation.FetchBundle(ctx, td, url, s.cfg.SPIFFEFetchOptions...)
+		if err != nil {
+			return nil, trace.Wrap(err, "fetching bundle using https_web profile")
+		}
+
+		// Calculate the duration before we should next sync, applying any
+		// sensible upper and lower bounds.
+		nextSyncIn = s.cfg.DefaultSyncInterval
+		if refreshHint, ok := bundle.RefreshHint(); ok {
+			if refreshHint < s.cfg.MinSyncInterval {
+				log.InfoContext(
+					ctx,
+					"Refresh hint is less than MinSyncInterval, using MinSyncInterval",
```

Comment:
Very much a nitpick, but I think if I saw this message logged I'd be curious what `MinSyncInterval` is. Maybe worth including it as another parameter? (And possibly same for `MaxSyncInterval` below).

I'd also assume if this gets logged once, it'll probably get logged... well, every minute. It might get a bit spammy, though I'm not sure I have a better idea, other than _maybe_ reducing it to debug level. Maybe not worth worrying about 🤷

Reply:
Good catch!

Reply:
6a40f8207b